staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go

```diff
@@ -248,17 +267,27 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
 	if err := validateEncryptedObject(request); err != nil {
 		return nil, err
 	}
-	return proto.Marshal(request)
+
+	buffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
+	defer t.encryptedObjectDecoder.bufferPool.Put(buffer)
+
+	if err := buffer.Marshal(request); err != nil {
+		return nil, err
+	}
+
+	return buffer.Bytes(), nil
 }
 
 // doDecode decodes the byte array to an EncryptedObject.
 func (t *envelopeTransformer) doDecode(originalData []byte) (*kmstypes.EncryptedObject, error) {
-	o := &kmstypes.EncryptedObject{}
-	if err := proto.Unmarshal(originalData, o); err != nil {
-		return nil, err
-	}
-	// validate the EncryptedObject
-	if err := validateEncryptedObject(o); err != nil {
+	buffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
+	defer t.encryptedObjectDecoder.bufferPool.Put(buffer)
+	buffer.SetBuf(originalData)
+
+	o := t.encryptedObjectDecoder.objectPool.Get().(*kmstypes.EncryptedObject)
+	defer t.encryptedObjectDecoder.objectPool.Put(o)
+
+	if err := buffer.Unmarshal(o); err != nil {
```

Comment:
What do you expect this to do? This does basically the same what we're already doing, with addtional complexity:

https://github.com/kubernetes/kubernetes/blob/4954c7bac4029d2f2e4b305fdba41f81b718aefc/vendor/github.com/gogo/protobuf/proto/decode.go#L396-L402

vs

https://github.com/kubernetes/kubernetes/blob/4954c7bac4029d2f2e4b305fdba41f81b718aefc/vendor/github.com/gogo/protobuf/proto/decode.go#L334-L338

Reply:
See https://github.com/kubernetes/kubernetes/pull/119167#discussion_r1260181086